src/mc-fle2-insert-update-payload-v2.c

```diff
@@ -348,8 +401,117 @@ bool mc_FLE2InsertUpdatePayloadV2_serializeForRange(const mc_FLE2InsertUpdatePay
+
 #undef IUPS_APPEND_BINDATA
 
+#define SERIALIZE_ARRAY_OF_TEXT_TOKEN_SET_FOR_TYPE_IMPL(Type)                                                          \
+    static bool _fle2_serialize_array_of_Text##Type##TokenSet(bson_t *parent,                                          \
+                                                              const char *array_name,                                  \
+                                                              const mc_array_t *array) {                               \
+        BSON_ASSERT_PARAM(parent);                                                                                     \
+        BSON_ASSERT_PARAM(array_name);                                                                                 \
+        BSON_ASSERT_PARAM(array);                                                                                      \
+        bson_t arr_bson;                                                                                               \
+        if (!BSON_APPEND_ARRAY_BEGIN(parent, array_name, &arr_bson)) {                                                 \
+            return false;                                                                                              \
+        }                                                                                                              \
+                                                                                                                       \
+        const char *index_string = NULL;                                                                               \
+        char storage[16];                                                                                              \
+        uint32_t index = 0;                                                                                            \
+                                                                                                                       \
+        for (size_t i = 0; i < array->len; i++) {                                                                      \
+            mc_Text##Type##TokenSet_t ts = _mc_array_index(array, mc_Text##Type##TokenSet_t, i);                       \
+            bson_uint32_to_string(index, &index_string, storage, sizeof(storage));                                     \
+                                                                                                                       \
+            if (!_fle2_serialize_Text##Type##TokenSet(&arr_bson, index_string, &ts)) {                                 \
+                return false;                                                                                          \
+            }                                                                                                          \
+            if (index == UINT32_MAX) {                                                                                 \
```

Comment:
Can we just iterate up to `BSON_MIN(array->len, UINT32_MAX)` instead of doing this check every iteration?

Reply:
done